core/editor.js

```diff
@@ -328,4 +338,84 @@ function normalizeDelta(delta) {
   }, new Delta());
 }
 
+function splitDelta(delta, index) {
+  return [delta.slice(0, index), delta.slice(index)];
+}
+
+function diffDeltas(oldDelta, newDelta, selectionInfo = undefined) {
+  if (selectionInfo) {
```

Comment:
Since there's no substantial else branch we can we instead check the opposite and return right away and save an indent on a lot of code? Also the inner if + else if both check for the same second condition which could be made more obvious by checking outside/immediately too so adding both of this together we could do something like:

```js
const [oldSelection, newSelection] = selectionInfo || [];
if (selectionInfo == null || newSelection.length > 0) {
  return oldDelta.diff(newDelta);
}
```

Reply:
The branches of the big if/else don't always return; sometimes they fall through.  So the diff has to be at the bottom to catch those cases.

The way I think of this method is as two special cases to check for, each with its own conditions, checks, and guards.  It's possible that the number of special cases or the conditions and guards on each case could change in the future.  Maybe this would be clearer if the big if/else were actually two ifs.  `selectionInfo` has to be present to even consider the special cases, and to calculate the variables that are used by both.  There's no way to know if a special case will apply or not, though, without running them.

Reply:
I don't feel too strongly, but is duplicating code better than having an extra level of indentation?  If the non-special case needs to change, it would have to change in two places.

The fact that the two special cases (insertion or deletion before or after the cursor; typing over a range) both make certain requirements of newSelection and oldSelection, and these requirements are partly the same and partly exclusive, seems totally coincidental to me, so at least the way I've been thinking about it, highlighting the common part and the exclusive part doesn't add clarity.  But maybe just making the structure of the code simpler makes it easier to read.  We can do it.

Reply:
I mean maybe there is a logic to it, like if the new selection is a range then logically, this isn't a splice caused by someone typing.  And if the old selection is a range that's a different case than the old selection being a cursor, and the two diff strategies basically correspond to those two cases.  I guess I just didn't have those thoughts at the time. :)

Reply:
I don't feel strongly on my particular recommendation but I do feel *something* should be done to make the if branch easier to read. At 70 lines long I cannot see the full if branch on a 28" monitor. A couple other ideas:

- The two inner if else branches can be their own functions 
- Instead of calling `return oldDelta.diff(newDelta);` at the end we call `diffDeltas(oldDelta, newDelta)` (removing the selectionInfo parameter even if one is supplied)